### DISCUSSION TITLE
能不能支持Yew框架呢！

### DISCUSSION THREAD

**Discussion by lhjok** [ant-design/ant-design#37660] [Ideas]
现在的Yew已经很好用了，但是缺乏antd这样的组件库，不知道能不能支持一下，可以尽快普及Wasm的应用。每次切换一个语言来写代码，都有一段不适应期，能用一个语言的完成全栈开发，那不是一件挺爽的事情吗！

**Comment by SignDawn** [comment]
js 不是也可以用一个语言来完成全栈开发？

**Reply by lhjok** [reply]
你说的也没错，Js没有Rust通用，我用Rust还有其它用处。Rust是我比较喜欢的语言，也是原因之一。

**Comment by fnoopv** [comment]
这个问题应该框架去支持，我看了一下官网，说是React开发没问题，但是官方没有给出例子，怎么结合，怎么通信。让一个UI库去适配它的HTML语法成本太高，很难的

**Reply by lhjok** [reply]
框架支持不太现实。确实需要使用Rust重新实现一遍Antd，但大部分设计不需要改变，Yew和Rust很相似的。

**Reply by lhjok** [reply]
Antd不是也支持Vue吗？相比而言，Yew和React更像，并且使用最新的技术，长期来看Wasm前途会更广阔。

**Reply by fnoopv** [reply]
> 框架支持不太现实。确实需要使用Rust重新实现一遍Antd，但大部分设计不需要改变，Yew和Rust很相似的。

框架直接给出怎么使用react的实例即可，react能使用就能使用antd

**Comment by MadCcc** [comment]
目前 antd-vue 和 antd-angular 都是社区实现。如果有兴趣的话欢迎贡献～

**Comment by jasper2007111** [comment]
实现Ant的Vue只是同一种语言的不同框架的迁移，难度相对不是很大，可以在操作上保留Ant一致的方式，但是用Yew的话，实现出来的Ant可能无法保证跟原来组件一样的使用方式，虽然可以做到风格上的一致。但是因为JS跟Rust是两个完全不同的语言，JS很动态，Rust又非常严格。

**Reply by lhjok** [reply]
`wasm_bindgen`可以和js互通，使用上或许有些不便，但可以做到和Ant的体验相差不远。

**Reply by jasper2007111** [reply]
建议你亲自尝试写几个Yew的Ant组件，逻辑实现可以参考Ant官方的，等别人实现不如自己先试试这个想法了，反正大家都是程序员，我目前就是自己在尝试的写Element的Yew版组件，简单的组件问题不大，但是类似Table这样的组件，会发现语言不同导致的实现也是不一样的。

**Reply by lhjok** [reply]
我现在对`Dioxus`和`Leptos`比较感兴趣，也尝试基于开源的`Tailwind Elements`组件库，但碰到很多问题。如果你想偷懒，那么融入第三方JS库，会让你很抓狂。这些`Rust`前端框架几乎都无法处理第三方JS库的事件监听器，只能由第三方库同时支持初始化方法和事件分离的方法。因为`Rust`前端框架，在卸载组件时并不会自动清除事件监听器，而且即使你在父组件注册了事件监听器，子组件一但切换，就中断了事件监听器的联系。我已经反馈了这些问题，`Tailwind Elements`组件库开发人员声明并不会支持我的想法，`Rust`前端框架则表现的很无奈，似乎很难解决这个问题。

**Reply by lhjok** [reply]
由于无法直接使用全局事件监听器，所以必须要在组件内注册事件监听器，然后在组件卸载后将其事件监听器清除。但第三方库不提供方法的话，是无法手动清除事件监听器的，这样会导致重复注册事件监听器，并出现内存泄漏，或特效重叠的问题。`Rust`前端框架全部都无法支持全局事件监听器，这可能是因为卸载组件重新再装载后，他们的对象是不一样的，所以绑定原对象的事件监听器会失去效果。

**Reply by lhjok** [reply]
这些仅发生在第三方JS库，如果你自己造轮子，那不会有任何问题。

**Reply by jasper2007111** [reply]
嗯，我目前就是自己造轮子了，你也造轮子吧，等待别人的东西不如自己弄，这也是学习Rust一个很好的机会，也是学习Web前端的好机会。
